### DISCUSSION TITLE
Wanted: Powershell script example of new 2021 auth?

### DISCUSSION THREAD

**Discussion by insaneoctane** [timdorr/tesla-api#303] [Q&A]
My coding experience is limited and I tend to stand on the shoulders of greatness.... I don't have the chops to write these headers, cookies, POSTS, SHA-256 hashes and stuff from scratch!  So, I'm calling on someone will real skills to help turn Tim's work and maybe others progress into a powershell script that I can use to fix my powershell automation script.  Thanks.

**Comment by insaneoctane** [comment]
Bump.  My current token expires in about 6 days :( 
Pretty please?

**Reply by IMgoRt** [reply]
Although not a direct answer, I used this https://github.com/fredli74/fetch-tesla-token to generate a new token (the "old" refresh_token you have won't work).
Once you have that then the code for step 3 and 4 is straight forward (mine's in NodeRed), basically:
1. you use the refresh_token as input into step 3, which creates a bearer_token
2. you then [within 300s] take the bearer_token as input into step 4 which produces the new token

(this is only needed within 45 days as the token expires, it is said that the refresh_token never expires, but I haven't tested that.

**note there appears to be a live issue with the auth process readers on the tesla servers so fetch-tesla-token may fail right now

**Comment by MrWugz** [comment]
I've written a [proof of concept PowerShell script](https://github.com/MrWugz/Get-TeslaToken) for retrieving the API token.  It doesn't support MFA yet, but does support refreshing the API token using the existing OAuth token.  Both are saved as .xml objects to the directory where the script resides.

New to GitHub, so feedback is appreciated.
